ecosystem/sep-0039.md

```diff
@@ -0,0 +1,193 @@
+
+
+## Specification
+First, we will discuss a common NFT issuance model and its ownership implications. This is not a required model, but sets the stage for why a standardized encoding of digital assets is necessary.
+
+### NFT Issuance
+Given an arbitrary digital asset, here's how issuance typically works:
+
+  1. An account is created to represent the digital asset on the ledger
+  2. The account stores the entire digital asset within its managed data entries.
+  3. The account issues a Stellar asset representing ownership over the digital asset we just stored.
+  4. Any account holding that Stellar asset lays claim to the digital asset.
+
+Notice that ownership of the token directly conveys ownership of a unit of the asset, entirely within the Stellar ledger. Other scenarios such as locking down the issuing account to avoid supply increases, ownership transfers via payments, etc. are built on top of this basic model and are thus extraneous to this SEP.
+
+Our focus is Step 2: storing the asset within the account.
+
+### Encoding Binary Data
+We begin by describing how binary data should be represented in a `ManageData` entry.
+
+Since the _key_ is restricted specifically to printable characters (ref: Stellar Core's [`isString32Valid()`](https://github.com/stellar/stellar-core/blob/b031954458df3bb31d8d98f136c6fee40523a10d/src/util/types.cpp#L89-L101) and C's [`isprint()`](https://en.cppreference.com/w/cpp/string/byte/isprint)), this gives us an encoding space of 94 characters or ~6.5 bits per character (a ~25% reduction in the space). From some cursory research, the [BasE91] encoding scheme closest to this theoretical limit and outperforms base64 encoding.
+
+The _value_ can be any binary data; thus, no encoding is necessary.
+
+### Representing the Data
+With the encoding described [above](#encoding-binary-data), we can now discuss the actual storage. Again, we'll use the `ManageData` operation to store the binary data. 
+
+The first entry _may_ be a [MIME type], followed by the first chunk of data, separated by a comma character (`,`). The MIME type *must* fit in a single key, and so must be <= 63 characters long.
+
+If the MIME type is excluded, the comma _must_ still be included so that readers can easily distinguish where the type starts and the data begins.
+
+Every subsequent entry is simply the next BasE91-encoded chunk, sized to fit into the 64-length key string, and the next 64-byte binary chunk.
+
+If the final entry fits into the key (that is, the value is not necessary), the encoder _must_ split it into a key and a value. Specifically, since the number of bytes remaining is definitely <= 64, the key _must_ be composed of dash (`-`) characters and the value _must_ be the final chunk.
+
+For example,
+
+| Key | Value |
+|:----|:-----:|
+| text/plain,'ZP^gp@..<etc>.. | .. 64 bytes .. |
+| Io1Tc%ZEYkE#^IR`0e..<etc>.. | .. 64 bytes .. |
+| %yO){B&mA#_1:W8{lT..<etc>.. | .. 64 bytes .. |
+| Cvnu8jZBP>1]TX2$g%..<etc>.. | .. 64 bytes .. |
```

Comment:
This is a really good point actually 🤔

Reply:
Instead of using a mandatory index prefix, could we not leave the generation of unique keys to the implementation? A different size `n` (e.g. 63 second iteration) would likely break duplicate keys and would still overall be more efficient than storing an index for every key?

Reply:
Using a single byte vs two would be nice but at the end of the day we still need indexing in order to reassemble the keys in the appropriate order when coming from Horizon. So unless the Horizon API changes indexing for actual reassembly is necessary. In my view we implement the SEP as Horizon stands today and if it changes in the future we can update the SEP. What we may want to include is a version key in the metadata in order to hint to the client how to interpret the data in case it changes in the future either via Horizon or CAP updates

Reply:
I don't think Horizon can, nor should we expect it to, guarantee that the key-value pairs in a JSON object are ordered. [RFC7159 Section 1](https://datatracker.ietf.org/doc/html/rfc7159#section-1) states that a JSON object is an unordered collection of name-value pairs.

Reply:
A strong argument for prefixes for sure (especially in light of Tyler's other comment re: Horizon), but I would counter-argue that if your data has replicating keys, you are wasting space and should compress your data, instead. However, this is a subtlety that should at least be _addressed_ by the SEP, if not solved.

Again, though, it does lend further credence to a distinct, per-entry prefix.

Reply:
(I'm avoiding making definitive statements regarding the index prefix because I'm actively trying to determine what makes Horizon return things in order, first.)

Reply:
Here is one potentially radical question.

The system is designed for optimizing storage space and costs at the expense of being intuitive design-wise (using keys to store values). I am wondering if practically, it really makes a difference for users that the storage space is 50 kb or 100 kb regardless with such high storage costs  i.e. > 400 XLM is required in both cases. Would they not likely switch to off-chain storage anyways in either scenario (likely at much lower storage requirements).

And therefore, my question is: could we not simply keep using keys for keys and values for values to keep the system consistent design-wise and have less constraints for key naming? The space would be reduced by half but does not seem realistic to even get close to such limits due to costs anyways.

Reply:
The Go stdlib orders keys using this definition:

> The map keys are sorted and used as JSON object keys by applying the following rules, subject to the UTF-8 coercion described for string values above:
>
> - keys of any string type are used directly
> - encoding.TextMarshalers are marshaled
> - integer keys are converted to strings

Ref: https://pkg.go.dev/encoding/json#Marshal

Reply:
>could we not simply keep using keys for keys and values for values to keep the system consistent design-wise and have less constraints for key naming?

💯 I think there's a lot of value in this. In the near future we (@Shaptic or I) plan to propose a CAP that will make any attempts to place data in the key pointless anyway by allowing the value field to expand to use any unused key space.

If we go with this slightly less space efficient approach, the SEP is so much simpler to implement.

Reply:
> And therefore, my question is: could we not simply keep using keys for keys and values for values to keep the system consistent design-wise and have less constraints for key naming?

Short answer no. I think we should support a hyper optimized standard for arbitrary data storage on-chain. I for one will opt for higher cost for on-chain storage whenever it makes sense. I don't like IPFS. A lot of the data I store is pixel art so it tends to be incredibly small already. However there is a world of difference in the Colorglyph use case when you're talking about 100 XLM vs 20 XLM when scaled to tens of thousands of glyphs. It's a difference of affordable vs not. Reliable permeant storage isn't cheap, and it's paramount we make it as cheap as possible. 

Ideally we'd get a CAP that allows for the key:value paradigm to become more flexible so this goofiness is obfuscated away and it's more a simple 0.5 XLM per 128 bytes of data calculation. Until then though I think it's worth it to push for a standard to prove why it would be wise to implement a CAP. Maybe there too we can address the indexing issue and allow for a more raw storage mechanic billed at a rate of 0.5 XLM per 128 bytes and do away with key:value all together. You pick how you want to store and surface data. Ultimately we just need to push some standard and SQ and CG have explored many various on-chain standards and some form of what I've laid out here: https://github.com/stellar/stellar-protocol/pull/1090#issuecomment-983029317 is what currently makes the most sense to me.

Reply:
> 💯 I think there's a lot of value in this.

In short if this SEP goes this route my vote is we abandon it. I won't use it. It'll be too expensive. I'm fine pausing SEP work until we have more clarity around CAP and/or Horizon changes but I still think there's value in formalizing a standard and using that to drive deeper improvements at more foundational layers. The point here never was to use keys and values the way they were intended to but to cram the absolute max amount of usable data into an account as a permanent data store. Sacrificing some efficiency for standards around mime-type hints and additional metadata fields is fine. Losing 64 bytes per entry entirely negates the reason we need standards in the first place and we'll revert back to everyone just smushing max data into accounts however it fits their singular isolated use case. This is first about saving money while maintaining on-chain storage, then about intended use case standards.

Reply:
> Ideally we'd get a CAP that allows for the key:value paradigm to become more flexible so this goofiness is obfuscated away and it's more a simple 0.5 XLM per 128 bytes of data calculation. Until then though I think it's worth it to push for a standard to prove why it would be wise to implement a CAP.

The CAP would be an ideal solution in the future. As I understand it would still require a SEP to define the nomenclature for keys so maybe the SEP here could also define the required key nomenclature so we can start adopting a more consistent approach with the keys optimization yet still have the flexibility of backward-compatibility for use cases like Colorglyph which require a working standard now.

Reply:
Lost this thread in the (awesome) back-and-forth we have here. Just chiming in on this:

>  I am wondering if practically, it really makes a difference for users that the storage space is 50 kb or 100 kb regardless with such high storage costs i.e. > 400 XLM is required in both cases. 

I'd say absolutely it does. A 2x size bump is the difference between Axie Infinities and Bored Apes ([article](https://medium.com/web-design-web-developer-magazine/at-what-pixel-dimensions-are-most-generative-nft-art-projects-built-9344d3998ac0), napkin math). The storage cost (around 150-200 USD at current XLM costs) is mildly prohibitive, but it also becomes part of the intrinsic value of the NFT itself.

Agreed on all counts that a CAP modifying `ManageData` would greatly simplify this SEP; perhaps in v2..